store/auth.go

```diff
@@ -166,3 +173,130 @@ func DischargeAuthCaveat(username, password, macaroon, otp string) (string, erro
+	var responseData struct {
+		Nonce string `json:"nonce"`
+	}
+	if err := dec.Decode(&responseData); err != nil {
+		return nil, fmt.Errorf(errorPrefix+"%v", err)
+	}
+
+	if responseData.Nonce == "" {
+		return nil, fmt.Errorf(errorPrefix + "no nonce returned")
+	}
+	return []byte(responseData.Nonce), nil
+}
+
+func requestDeviceMacaroon(serialAssertion []byte, nonce []byte, signature []byte) (string, error) {
+	const errorPrefix = "cannot get device session macaroon from store: "
+
+	type deviceMacaroonRequest struct {
+		SerialAssertion string `json:"serial-assertion"`
+		Nonce           string `json:"nonce"`
+		Signature       string `json:"signature"`
+	}
+
+	jsonData, err := json.Marshal(deviceMacaroonRequest{
+		SerialAssertion: string(serialAssertion),
+		Nonce:           string(nonce),
+		Signature:       string(signature),
+	})
+
+	req, err := http.NewRequest("POST", DeviceIdentityAPI+"/sessions", bytes.NewBuffer([]byte(jsonData)))
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		// TODO: We should return a more useful error message if
+		// possible. The store's device macaroon APIs produce
+		// error objects in a different format from the user
+		// macaroon APIs, so they should probably be assimilated
+		// first.
+		return "", fmt.Errorf(errorPrefix+"store server returned status %d", resp.StatusCode)
+	}
+
+	dec := json.NewDecoder(resp.Body)
+	var responseData struct {
+		Macaroon string `json:"macaroon"`
+	}
+	if err := dec.Decode(&responseData); err != nil {
+		return "", fmt.Errorf(errorPrefix+"%v", err)
+	}
+
+	if responseData.Macaroon == "" {
+		return "", fmt.Errorf(errorPrefix + "no macaroon returned")
+	}
+	return responseData.Macaroon, nil
+}
+
+type signNonceFunc func(deviceKeyEncoded []byte, nonce []byte) ([]byte, error)
+
+// AcquireDeviceMacaroon obtains a macaroon for a device session by requesting and signing a nonce.
+func AcquireDeviceMacaroon(serialAssertionEncoded []byte, signNonceFunc signNonceFunc) (string, error) {
+	const errorPrefix = "cannot get device session macaroon from store: "
+
+	// Extract the public key from the assertion, to pass to the
+	// nonce signing function.
+	assert, err := asserts.Decode(serialAssertionEncoded)
+	if err != nil {
+		return "", err
+	}
+	if assert.Type() != asserts.SerialType {
+		return "", fmt.Errorf("serial assertion is actually %s", assert.Type())
+	}
+	serial := assert.(*asserts.Serial)
+	deviceKeyEncoded, err := asserts.EncodePublicKey(serial.DeviceKey())
+	if err != nil {
+		return "", err
+	}
+
+	// Get a nonce from the store, sign it, and then exchange it for
+	// a macaroon.
+	nonce, err := requestDeviceNonce()
+	if err != nil {
+		return "", err
+	}
+
+	signature, err := signNonceFunc(deviceKeyEncoded, nonce)
```

Comment:
atm the serial assertions carry just a public key blob (and don't have explicit keyid or fingerprint headers), so in the end this is about the gadget and the store needing to agree, and snapd in between those,

for account-keys atm we are using prefix less fingeprints and keyids:
* we can introduce prefixes (because these are fixed size hex numbers, in a backward compatible ways if we discover we need to do)
* medium term we expect brands and developers to actually use GPG/PGP keys, so having fingeprints and keyids that look just like those seems friendlier
* PGP public key fingerprints (for the main algorithms) are relatively straightforward to compute even for "foreign keys", the only extra piece of metadata is a creation time (that can be set to 0 if it's really not available/meaningful) the rest is just bits from the public key and trivial formatting, so we could use PGP fingerprints even for bare/non-PGP keys